python/fusion_engine_client/utils/construct_utils.py

```diff
@@ -1,87 +1,126 @@
-    # skip _io member
-    return self.tuple_cls(*list(obj.values())[1:])
+    def __init__(self, tuple_cls, *args):
+        """!
+        @brief Create an adapter for (de)serializing NamedTuples.
+
+        @param tuple_cls The NamedTuple to adapt.
+        """
+        super().__init__(*args)
+        self.tuple_cls = tuple_cls
 
-  def _encode(self, obj, context, path):
-    return obj._asdict()
+    def _decode(self, obj, context, path):
+        # skip _io member
+        return self.tuple_cls(*list(obj.values())[1:])
+
+    def _encode(self, obj, context, path):
+        return obj._asdict()
 
 
 class ClassAdapter(Adapter):
-  """!
-  @brief Adapter for automatically converting between construct streams and
-         a class with corresponding fields.
-
-  Usage Example:
-  ```{.py}
-      class VersionClass:
-        def __init__(self, major=0, minor=0):
-          self.major = major
-          self.minor = minor
-
-      VersionRawConstruct = Struct(
-          "major" / Int8ul,
-          "minor" / Int16ul,
-      )
-
-      VersionConstruct = ClassAdapter(VersionClass, VersionRawConstruct)
-      UserConfigConstruct = Struct(
-          "version" / VersionConstruct,
-          "thing2" / Int32ul,
-      )
-      UserConfigConstruct.build({'version': VersionClass(2, 3), 'thing2': 4})
-  ```
-  """
-
-  def __init__(self, cls, *args):
     """!
-    @brief Create an adapter for (de)serializing a class.
+    @brief Adapter for automatically converting between construct streams and
+           a class with corresponding fields.
+
+    Usage Example:
+    ```{.py}
+        class VersionClass:
+            def __init__(self, major=0, minor=0):
+                self.major = major
+                self.minor = minor
+
+        VersionRawConstruct = Struct(
+            "major" / Int8ul,
+            "minor" / Int16ul,
+        )
+
+        VersionConstruct = ClassAdapter(VersionClass, VersionRawConstruct)
+        UserConfigConstruct = Struct(
+            "version" / VersionConstruct,
+            "thing2" / Int32ul,
+        )
+        UserConfigConstruct.build({'version': VersionClass(2, 3), 'thing2': 4})
+    ```
+    """
+
+    def __init__(self, cls, *args):
+        """!
+        @brief Create an adapter for (de)serializing a class.
+
+        @param cls The class to adapt.
+        """
+        super().__init__(*args)
+        self.cls = cls
+
+    def _decode(self, obj, context, path):
+        val = self.cls()
+        val.__dict__.update(obj)
+        return val
 
-    @param cls The class to adapt.
+    def _encode(self, obj, context, path):
+        return obj.__dict__
+
+
+class EnumAdapter(Adapter):
+    """!
+    @brief Adapter for automatically converting between construct Enum and
+           python Enums.
+
+    Usage Example:
+    ```{.py}
+        class ConfigType(IntEnum):
+            FOO = 0
+            BAR = 1
+
+        UserConfigConstruct = Struct(
+            "config_type" / Enum(Int16ul, ConfigType),
```

Comment:
Wait, `EnumAdapter` isn't used anywhere in this example? Should this be `EnumAdapter` and not `Enum`?